presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexFirstSet.java

```diff
@@ -0,0 +1,251 @@
+            Matcher m = Pattern.compile(pattern).matcher(hint);
+            checkState(m.find(), err);
+
+            int integer = Integer.parseInt(m.group(1));
+            String percent = m.group(2);
+
+            if (percent != null) {
+                if (integer < 1 || integer > MAX_VALUE_OF_INDEX_FIRST_MAX_PERCENT) {
+                    err = "Invalid 'threshold:${maxPercent}%' hint value '" + hintValue + "' of hint key '" + hintKey + "', "
+                            + "which[" + integer + "] should be within the range [1, " + MAX_VALUE_OF_INDEX_FIRST_MAX_PERCENT + "]";
+                    throw new IllegalArgumentException(err);
+                }
+                return thresholdWithPercent(integer);
+            }
+            else {
+                if (integer < 1 || integer > MAX_VALUE_OF_INDEX_FIRST_MAX_ROWS) {
+                    err = "Invalid 'threshold:${maxRows}' hint value '" + hintValue + "' of hint key '" + hintKey + "', "
+                            + "which[" + integer + "] should be within the range [1, " + MAX_VALUE_OF_INDEX_FIRST_MAX_ROWS + "]";
+                    throw new IllegalArgumentException(err);
+                }
+                return thresholdWithRows(integer);
+            }
+        }
+        if (hint.matches("^\\[[^\\[\\]]*]$")) {
+            hint = hint.substring(1, hint.length() - 1).trim();
+            if (StringUtils.isBlank(hint)) {
+                return IndexFirstSet.none();
+            }
+            String[] ss = hint.split("[ ;,]+");
+            Set<SchemaTableName> tables = Arrays.stream(ss)
+                    .filter(StringUtils::isNotBlank)
+                    .map(String::trim)
+                    .map(String::toLowerCase)
+                    .map(x -> checkAndAssemblySchemaTableName(currentConnectionSchema, hintKey, x))
+                    .collect(Collectors.toSet());
+            return IndexFirstSet.custom(tables);
+        }
+        String str = "use hint like these: " +
+                "1)use index if possible -> 'tablestore-index-first=auto' " +
+                "2)do not use any index, default -> 'tablestore-index-first=none' " +
+                "3)use indexes of tables that specified -> 'tablestore-index-first=[xxDb.yyTable, zzTable, ...]' " +
+                "4)use heuristic rule of max matched rows -> 'tablestore-index-first=threshold:1000' " +
+                "5)use heuristic rule of max matched percentage -> 'tablestore-index-first=threshold:5%'";
+        throw new IllegalArgumentException("Invalid hint value '" + hint + "' of hint key '" + hintKey + "', " + str);
+    }
+
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    private static SchemaTableName checkAndAssemblySchemaTableName(Optional<String> currentConnectionSchema,
+            String hintKey,
+            String unit)
+            throws IllegalArgumentException
```

Comment:
hmmm, why we need explicitly throws a runtime exception in a private method?

Reply:
You mean the exception should be throw in a public method?

Reply:
Not really, I'm just curious 
1. Why we need a "throws" for a RuntimeException?  We don't have to, am I right? (IllegalArgumentException is a runtime/unchecked exception, isn't it?)
2. In case you just want to use 'throws' to make a strong contract to the caller, but 'checkAndAssemblySchemaTableName' is a private method.  I cannot see the benefits of the 'throws' here.